lib/eco/export/jruby.py

```diff
@@ -0,0 +1,344 @@
+        return [self.method] + self.callsite_versions
+
+    def __str__(self):
+        return ('Method Version: %s id=%g called_from=[ %s ]' %
+                (self.method.name, self.id,
+                 ' '.join([str(cs) for cs in self.called_from])))
+
+
+class CallSite(object):
+
+    def __init__(self, id_, method, line):
+        self.id = int(id_)
+        self.method = method
+        self.line = int(line)
+        self.versions = []
+
+    def reachable(self):
+        return [self.method] + self.versions
+
+    def __str__(self):
+        return ('Callsite: %s id=%g line=%g' %
+                (self.method.name, self.id, self.line))
+
+
+class CallSiteVersion(object):
+
+    def __init__(self, id_, callsite, method_version):
+        self.id = int(id_)
+        self.callsite = callsite
+        self.method_version = method_version
+        self.calls = []
+
+    def reachable(self):
+        # Method_version isn't reachable - find it through calls
+        return [self.callsite] + self.calls
+
+    def __str__(self):
+        return ('Callsite Version: %s id=%g' %
+                (self.callsite.method.name, self.id))
+
+
+class Mega(object):
+
+    def __str__(self):
+        return 'Megamorphic'
+
+
+class JRubyExporter(object):
+    def __init__(self, tm):
+        self.tm = tm  # TreeManager object.
+        self.sl_functions = dict()
+        self._sl_output = list()
+        self._wrappers = list()
+        self._output = list()
+        self._sl_functions = list()
+
+    def export(self, path=None, run=False, profile=False):
+        if path is not None:
+            self._export_as_text(path)
+            return
+        if run:
+            return self._run()
+        elif profile:
+            return self._profile()
+
+    def _language_box(self, name, node):
+        if name == '<Ruby>':
+            self._walk_rb(node)
+
+    def _walk_rb(self, node):
+        while True:
+            node = node.next_term
+            sym = node.symbol
+            if isinstance(node, EOS):
+                break
+            if isinstance(sym, MagicTerminal):
+                self._language_box(sym.name, node.symbol.ast.children[0])
+            elif isinstance(sym, IndentationTerminal):
+                self._output.append(sym)
+            elif sym.name == '\r':
+                self._output.append('\n')
+            else:
+                self._output.append(sym.name)
+
+    def _export_as_text(self, path):
+        node = self.tm.lines[0].node # first node
+        self._walk_rb(node)
+        output = ''.join(self._output)
+        with open(path, 'w') as fp:
+            fp.write(''.join(output))
+
+    def _run(self):
+        f = tempfile.mkstemp(suffix='.rb')
+        settings = QSettings('softdev', 'Eco')
+        jruby_bin =str (settings.value('env_jruby').toString())
+        self._export_as_text(f[1])
+        # Run this command:
+        #     $ jruby -X+T FILE.rb
+        return subprocess.Popen([jruby_bin,
```

Comment:
Maybe the first three args could be put on one line to save a bit of vertical space?